### PR TITLE
Updated roll calls to newer specs

### PR DESCRIPTION
**Pull request by RaulinN**

### CODE REVIEW

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/RollCallHandler.scala [hunk 1]
```diff
@@ -27,15 +31,17 @@ case object RollCallHandler extends MessageHandler {
...
+  def handleCreateRollCall(rpcMessage: JsonRpcRequest): GraphMessage = {
+    val message: Message = rpcMessage.getParamsMessage.get
+
+    val f: Future[GraphMessage] = (dbActor ? DbActor.Write(rpcMessage.getParamsChannel, message)).map {
+      case DbActor.DbActorWriteAck => Left(rpcMessage)
+      case DbActor.DbActorNAck(code, description) => Right(PipelineError(code, description, rpcMessage.id))
+      case _ => Right(PipelineError(ErrorCodes.SERVER_ERROR.id, "Database actor returned an unknown answer", rpcMessage.id))
     }
+
+    Await.result(f, duration)
```
Comment (pierluca) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/RollCallHandler.scala:43]: is there any form of timeout on this await?

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/EventValidator.scala [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+trait EventValidator {
+  /**
+   * Generates a hash to validate event ids
+   *
+   * @param hash 2nd hash parameter
+   * @param timestamp 3rd hash parameter
+   * @param string 4th hash parameter
+   * @return the calculated id
+   */
+  def generateValidationId(hash: Hash, timestamp: Timestamp, string: String): Hash
```
Comment (pierluca) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/EventValidator.scala:14]: This abstraction seems to obfuscate things more than anything, as it doesn't really reduce the amount of code written and hides the link with the hash function.... I wouldn't mind reverting to a simple `Hash.fromStrings` as it was there before.

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala [hunk 3]
```diff
@@ -2,30 +2,34 @@ package ch.epfl.pop.pubsub.graph.validators
...
-          Right(validationError(s"the message should include either 'start' or 'scheduled' values, but not both"))
-        } else if (data.start.isDefined && !validateTimestampOrder(data.creation, data.start.get)) {
-          Right(validationError(s"'start' (${data.start.get}) timestamp is younger than 'creation' (${data.creation})"))
-        } else if (data.scheduled.isDefined && !validateTimestampOrder(data.creation, data.scheduled.get)) {
-          Right(validationError(s"'scheduled' (${data.scheduled.get}) timestamp is younger than 'creation' (${data.creation})"))
-        } else if (expectedHash != data.id) {
+        } else if (!validateTimestampOrder(data.creation, data.proposed_start)) {
+          Right(validationError(s"'proposed_start' (${data.proposed_start}) timestamp is younger than 'creation' (${data.creation})"))
+        } else if (!validateTimestampOrder(data.proposed_start, data.proposed_end)) {
+          Right(validationError(s"'proposed_end' (${data.proposed_end}) timestamp is younger than 'proposed_start' (${data.proposed_start})"))
```
Comment (pierluca) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala:31]: `earlier than` is better than `younger than`

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala [hunk 4]
```diff
@@ -34,38 +38,33 @@ case object RollCallValidator extends MessageDataContentValidator {
...
 
     rpcMessage.getParamsMessage match {
       case Some(message: Message) =>
         val data: OpenRollCall = message.decodedData.get.asInstanceOf[OpenRollCall]
 
-        if (!validateTimestampStaleness(data.start)) {
-          Right(validationError(s"stale 'start' timestamp (${data.start})"))
+        val laoId: Hash = rpcMessage.extractLaoId
+        // Note: this id doesnt follow the standard sh256(string, hash, timestamp, string) other events do;
+        // this is why generateValidationId cannot be used
```
Comment (pierluca) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala:50]: If you need to write such a comment, it's one more reason to drop the `generateValidationId` and just use Hash everywhere ;-)

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala [hunk 5]
```diff
@@ -34,38 +38,33 @@ case object RollCallValidator extends MessageDataContentValidator {
...
-        val data: ReopenRollCall = message.decodedData.get.asInstanceOf[ReopenRollCall]
-
-        if (!validateTimestampStaleness(data.start)) {
-          Right(validationError(s"stale 'start' timestamp (${data.start})"))
-        } else {
-          Left(rpcMessage)
-        }
-      case _ => Right(validationErrorNoMessage(rpcMessage.id))
-    }
+    validateOpenRollCall(rpcMessage, "ReopenRollCall")
```
Comment (pierluca) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala:67]: why dropping the other checks?

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala [hunk 6]
```diff
@@ -34,38 +38,33 @@ case object RollCallValidator extends MessageDataContentValidator {
...
   }
 
   // TODO check that this is correct (correct validations)
```
Comment (pierluca) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala:70]: I'd consider dropping this todo :)

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala [hunk 7]
```diff
@@ -76,10 +75,19 @@ case object RollCallValidator extends MessageDataContentValidator {
...
+        val expectedRollCallId: Hash = Hash.fromStrings(
+          EVENT_HASH_PREFIX, laoId.toString, data.closes.toString, data.closed_at.toString
+        )
+
+        if (!validateTimestampStaleness(data.closed_at)) {
+          Right(validationError(s"stale 'closed_at' timestamp (${data.closed_at})"))
         } else if (data.attendees.size != data.attendees.toSet.size) {
           Right(validationError("duplicate attendees keys"))
+        } else if (expectedRollCallId != data.closes) {
+          Right(validationError("unexpected id 'closes'"))
```
Comment (pierluca) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala:90]: We're comparing the wrong item here.
the `closes` id should be in the database, but it's `update_id` which should be `== expectedRollCallId`
cf. https://github.com/dedis/student_21_pop/blob/master/fe1-web/model/network/method/message/data/rollCall/CloseRollCall.ts

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala [hunk 8]
```diff
@@ -34,38 +38,33 @@ case object RollCallValidator extends MessageDataContentValidator {
...
+        val laoId: Hash = rpcMessage.extractLaoId
+        // Note: this id doesnt follow the standard sh256(string, hash, timestamp, string) other events do;
+        // this is why generateValidationId cannot be used
+        val expectedRollCallId: Hash = Hash.fromStrings(
+          EVENT_HASH_PREFIX, laoId.toString, data.opens.toString, data.opened_at.toString
+        )
+
+        if (!validateTimestampStaleness(data.opened_at)) {
+          Right(validationError(s"stale 'opened_at' timestamp (${data.opened_at})"))
+        } else if (expectedRollCallId != data.opens) {
```
Comment (pierluca) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala:57]: We're comparing the wrong item here.
the `opens` id should be in the database, but it's `update_id` which should be `== expectedRollCallId`
cf. https://github.com/dedis/student_21_pop/blob/master/fe1-web/model/network/method/message/data/rollCall/OpenRollCall.ts

**Review by pierluca — CHANGES_REQUESTED** [review]
A few bugs need addressing, but nice progress 👍

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/RollCallHandler.scala (hunk 1 above)
Comment (RaulinN) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/RollCallHandler.scala:43]: The future fails if `duration` is exceeded :)

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala (hunk 5 above)
Comment (RaulinN) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala:67]: The checks are still there. They are the same as openRollCall. Thus, ReopenRollCall calls the validator of openRollCall (line 67)

Comment (pierluca) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala:67]: Oh, good. Same thing I ended up doing in TypeScript 😅

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/EventValidator.scala (hunk 2 above)
Comment (RaulinN) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/EventValidator.scala:14]: I thought about it for a while and hesitated putting it into the code. Looking at it once again, i would tend to agree with you ^^ Will remove it

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala (hunk 7 above)
Comment (RaulinN) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/RollCallValidator.scala:90]: Nice catch! ^^'

**Pull request merged**